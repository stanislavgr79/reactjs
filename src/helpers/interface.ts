export interface IMovie {
  [x: string]: any;
  id?: number;
  title: string;
  tagline?: string;
  vote_average?: number;
  vote_count?: number;
  release_date: string;
  movieUrl: string;
  genres: string[] | { label: string, value: string }[];
  poster_path?: string | undefined;
  overview: string;
  runtime: number | string | undefined;
}
