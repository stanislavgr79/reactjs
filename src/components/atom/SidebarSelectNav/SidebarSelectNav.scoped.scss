.sidebar-select-nav {
    padding-left: 15px;
    .side-sort-select {
        width: 120px;
        border: none;
        font-weight: 600;
        background: none;
        color: #FFFFFF;
        background-color: #232323;
        &>option {
            font-weight: 600;
        }
        &:hover {
            color: #F65261;
            cursor: pointer;
        }
        &:focus {
            outline: none;
        }
    }
}