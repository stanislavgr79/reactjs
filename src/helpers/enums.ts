/* eslint-disable no-unused-vars */

/**
 * Enum for common SelectGenres.
 * @readonly
 * @enum {{label: string, value: string}}
 */
export const selectGenres = Object.freeze([
  { label: 'Action', value: 'Action' },
  { label: 'Adventure', value: 'Adventure' },
  { label: 'Animation', value: 'Animation' },
  { label: 'Comedy', value: 'Comedy' },
  { label: 'Documentary', value: 'Documentary' },
  { label: 'Drama', value: 'Drama' },
  { label: 'Family', value: 'Family' },
  { label: 'Horror', value: 'Horror' },
  { label: 'Mystery', value: 'Mystery' },
  { label: 'Thriller', value: 'Thriller' },
]);

/**
 * Enum for common Genres.
 * @readonly
 * @enum { string }
 */
export enum Genres {
  ALL = '',
  DOCUMENTARY = 'documentary',
  COMEDY = 'comedy',
  HORROR = 'horror',
  CRIME = 'crime',
}

/**
 * Enum for common SortBy.
 * @readonly
 * @enum { string }
 */
export enum SortBy {
  TITLE = 'title',
  RELEASE = 'release_date',
  RUNTIME = 'runtime',
  RATING = 'vote_average',
}

/**
 * Enum for common SortOrder.
 * @readonly
 * @enum { string }
 */
export enum SortOrder {
  ASC = 'asc',
  DESC = 'desc',
}
