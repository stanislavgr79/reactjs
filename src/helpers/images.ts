const images = [
  {
    name: 'Ricky and Morty.jpg',
    value: './resources/images/Ricky and Morty.jpg',
  },
  {
    name: 'Pulp Fiction.jpg',
    value: './resources/images/Pulp Fiction.jpg',
  },
  {
    name: 'Kill Bill.jpg',
    value: './resources/images/Kill Bill.jpg',
  },
  {
    name: 'Avengers.jpg',
    value: './resources/images/Avengers.jpg',
  },
  {
    name: 'Inception.jpg',
    value: './resources/images/Inception.jpg',
  },
  {
    name: 'Reservoir Dogs.jpg',
    value: './resources/images/Reservoir Dogs.jpg',
  },
];

export default images;
