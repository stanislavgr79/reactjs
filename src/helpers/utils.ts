export const createClassList = (classesArray: string[]) =>
  classesArray
    .filter((className) => typeof className === 'string' && className.length > 0)
    .join(' ')
    .trim();

export function transformGenresToStringArray(genres: { value: string, label: string }[]) {
  const genresArray: string[] = [];
  genres.map((it) => genresArray.push(it.value));
  return genresArray;
}

export function transformGenresToArrayObjects(genres: string[]) {
  const genresArray: { value: string, label: string }[] = [];
  genres.map((it: string) =>
    genresArray.push({
      label: it,
      value: it,
    }),
  );
  return genresArray;
}
