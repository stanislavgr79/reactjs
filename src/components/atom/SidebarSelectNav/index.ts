import SidebarSelectNav from './SidebarSelectNav';

export default SidebarSelectNav;
