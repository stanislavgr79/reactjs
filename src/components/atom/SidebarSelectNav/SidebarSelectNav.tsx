import React, { useCallback, useEffect, useState } from 'react';
import { SortBy } from '../../../helpers/enums';

import './SidebarSelectNav.scoped.scss';

interface IProps {
  defaultValue: SortBy;
  updateCurrentSortBy: (sortBy: SortBy) => void;
}

export default function SidebarSelectNav({
  defaultValue,
  updateCurrentSortBy,
}: IProps): JSX.Element {
  const [selectedValue, setSelectedValue] = useState(defaultValue);
  const sideSortBy = Object.entries(SortBy);

  const handleChange = useCallback((event: any) => {
    setSelectedValue(event.target.value);
  }, []);

  useEffect(() => {
    updateCurrentSortBy(selectedValue);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedValue]);

  return (
    <div className="sidebar-select-nav">
      <select className="side-sort-select" onChange={(e) => handleChange(e)} value={selectedValue}>
        {sideSortBy.map(([key, value], index: number) => (
          <option
            id={`${value}-${index}`}
            value={value}
            key={index}
            className="c-pointer"
            role="button"
          >
            {key}
          </option>
        ))}
      </select>
    </div>
  );
}
