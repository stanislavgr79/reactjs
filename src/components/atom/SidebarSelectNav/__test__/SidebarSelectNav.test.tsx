import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import storeReal from '../../../../redux/redux-store';

import { render, screen, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SidebarSelectNav from '../SidebarSelectNav';
import { SortBy } from '../../../../helpers/enums';

describe('SidebarSelectNav', () => {
  const initialState = { sidebar: { selectedIndexListNav: 0 } };
  const mockStore = configureStore();
  let container: Element 
    | DocumentFragment 
    // eslint-disable-next-line prettier/prettier
    | RenderOptions<typeof import('@testing-library/dom/types/queries'), HTMLElement>
    | null;

  const updateCurrentSortByMock = jest.fn();
  
  afterEach(() => {
    container = null;
    jest.clearAllMocks();
  });

  test('render', async () => {
    const store = mockStore(initialState);

    container = render(
      <Provider store={store}>
          <SidebarSelectNav
            defaultValue={SortBy.TITLE}
            updateCurrentSortBy={updateCurrentSortByMock}
          />
      </Provider>,
    );
    
    expect(container.baseElement).toMatchSnapshot();
  });

  test('render SortBy.RATING select', async () => {
    const store = storeReal;

    container = render(
      <Provider store={store}>
          <SidebarSelectNav
            defaultValue={SortBy.TITLE}
            updateCurrentSortBy={updateCurrentSortByMock}
          />
      </Provider>,
    );

    expect(container.baseElement).toMatchSnapshot();
    userEvent.selectOptions(screen.getByRole('combobox', { name: '' }), ['runtime']);
    expect(container.baseElement).toMatchSnapshot();
    expect(container.baseElement)
      .toContainHTML('<select class="sidebar-select-nav selected-runtime">');
  });

  test('handleChange function', async () => {
    const store = mockStore(initialState);

    render(
      <Provider store={store}>
          <SidebarSelectNav
            defaultValue={SortBy.TITLE}
            updateCurrentSortBy={updateCurrentSortByMock}
          />
      </Provider>,
    );
    
    userEvent.selectOptions(screen.getByRole('combobox', { name: '' }), ['runtime']);
    expect(updateCurrentSortByMock).toBeCalled();
  });
});
