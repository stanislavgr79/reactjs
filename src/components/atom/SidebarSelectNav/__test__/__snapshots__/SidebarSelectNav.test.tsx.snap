// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarSelectNav render 1`] = `
<body>
  <div>
    <div
      class="sidebar-select-nav"
    >
      <select
        class="sidebar-select-nav selected-title"
      >
        <option
          class="c-pointer"
          id="title-0"
          role="button"
          value="title"
        >
          TITLE
        </option>
        <option
          class="c-pointer"
          id="release_date-1"
          role="button"
          value="release_date"
        >
          RELEASE
        </option>
        <option
          class="c-pointer"
          id="runtime-2"
          role="button"
          value="runtime"
        >
          RUNTIME
        </option>
        <option
          class="c-pointer"
          id="vote_average-3"
          role="button"
          value="vote_average"
        >
          RATING
        </option>
      </select>
    </div>
  </div>
</body>
`;

exports[`SidebarSelectNav render SortBy.RATING select 1`] = `
<body>
  <div>
    <div
      class="sidebar-select-nav"
    >
      <select
        class="sidebar-select-nav selected-title"
      >
        <option
          class="c-pointer"
          id="title-0"
          role="button"
          value="title"
        >
          TITLE
        </option>
        <option
          class="c-pointer"
          id="release_date-1"
          role="button"
          value="release_date"
        >
          RELEASE
        </option>
        <option
          class="c-pointer"
          id="runtime-2"
          role="button"
          value="runtime"
        >
          RUNTIME
        </option>
        <option
          class="c-pointer"
          id="vote_average-3"
          role="button"
          value="vote_average"
        >
          RATING
        </option>
      </select>
    </div>
  </div>
</body>
`;

exports[`SidebarSelectNav render SortBy.RATING select 2`] = `
<body>
  <div>
    <div
      class="sidebar-select-nav"
    >
      <select
        class="sidebar-select-nav selected-runtime"
      >
        <option
          class="c-pointer"
          id="title-0"
          role="button"
          value="title"
        >
          TITLE
        </option>
        <option
          class="c-pointer"
          id="release_date-1"
          role="button"
          value="release_date"
        >
          RELEASE
        </option>
        <option
          class="c-pointer"
          id="runtime-2"
          role="button"
          value="runtime"
        >
          RUNTIME
        </option>
        <option
          class="c-pointer"
          id="vote_average-3"
          role="button"
          value="vote_average"
        >
          RATING
        </option>
      </select>
    </div>
  </div>
</body>
`;
